ur_controllers/src/speed_scaling_state_controller.cpp

```diff
@@ -19,10 +19,187 @@
+  {
+    RCLCPP_INFO(get_node()->get_logger(), "Parameter 'state_publish_rate' not set");
+    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
+  }
+  else
+  {
+    RCLCPP_INFO(get_node()->get_logger(), "Publisher rate set to : %.1f Hz", publish_rate_);
+  }
+
+  try
+  {
+    speed_scaling_state_publisher_ =
+        get_node()->create_publisher<std_msgs::msg::Float64>("speed_scaling_factor", rclcpp::SystemDefaultsQoS());
+  }
+  catch (const std::exception& e)
+  {
+    // get_node() may throw, logging raw here
+    fprintf(stderr, "Exception thrown during init stage with message: %s \n", e.what());
+    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
+  }
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SpeedScalingStateController::on_activate(const rclcpp_lifecycle::State& /*previous_state*/)
+{
+  if (!init_sensor_data())
+  {
+    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
+  }
+
+  last_publish_time_ = node_->now();
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SpeedScalingStateController::on_deactivate(const rclcpp_lifecycle::State& /*previous_state*/)
+{
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+template <typename T>
+bool has_any_key(const std::unordered_map<std::string, T>& map, const std::vector<std::string>& keys)
+{
+  bool found_key = false;
+  for (const auto& key_item : map)
+  {
+    const auto& key = key_item.first;
+    if (std::find(keys.cbegin(), keys.cend(), key) != keys.cend())
+    {
+      found_key = true;
+      break;
+    }
+  }
+  return found_key;
+}
+
+bool SpeedScalingStateController::init_sensor_data()
+{
+  // loop in reverse order, this maintains the order of values at retrieval time
+  for (auto si = state_interfaces_.crbegin(); si != state_interfaces_.crend(); si++)
+  {
+    // initialize map if name is new
+    if (name_if_value_mapping_.count(si->get_name()) == 0)
+    {
+      name_if_value_mapping_[si->get_name()] = {};
+    }
+    // add interface name
+    name_if_value_mapping_[si->get_name()][si->get_interface_name()] = kUninitializedValue;
+  }
+
+  // filter state interfaces that have a speed scaling factor
+  // the rest will be ignored for this message
+  for (const auto& name_ifv : name_if_value_mapping_)
+  {
+    const auto& interfaces_and_values = name_ifv.second;
+    if (has_any_key(interfaces_and_values, { "speed_scaling_factor" }))
+    {
+      sensor_names_.push_back(name_ifv.first);
+    }
+  }
+
+  return true;
+}
+
+double get_value(const std::unordered_map<std::string, std::unordered_map<std::string, double>>& map,
```

Comment:
Can be removed if only one state interface is configured. This copy paste from joint state controller would be handy if we had to work with all the interfaces in the system.

Reply:
Done